sources/packages/backend/apps/queue-consumers/src/processors/schedulers/institution-integration/ier12-integration/_tests_/e2e/ier12-factory.ts

```diff
@@ -0,0 +1,342 @@
+  db: E2EDataSources,
+  testInputApplication: IER12Application,
+  student: Student,
+  institutionLocation: InstitutionLocation,
+  programYear: ProgramYear,
+  createSecondDisbursement: boolean,
+  referenceSubmission: Date,
+): Promise<Application> {
+  const application = await saveFakeApplicationDisbursements(
+    db.dataSource,
+    { student, institutionLocation, disbursementValues: [] },
+    { createSecondDisbursement },
+  );
+  application.applicationNumber = testInputApplication.applicationNumber;
+  application.studentNumber = testInputApplication.studentNumber;
+  application.relationshipStatus = testInputApplication.relationshipStatus;
+  application.submittedDate =
+    testInputApplication.submittedDate ?? referenceSubmission;
+  application.applicationStatus = testInputApplication.applicationStatus;
+  application.applicationStatusUpdatedOn =
+    testInputApplication.applicationStatusUpdatedOn ?? referenceSubmission;
+  application.programYear = programYear;
+  await db.application.save(application);
+  return application;
+}
+
+/**
+ * Populates the assessments and its awards from the test input data
+ * and saves them all.
+ * @param db data source helper.
+ * @param testInputAssessment assessment and awards test input data.
+ * @param assessment previously saved assessment to be updated.
+ * @param studyStartDate offering start date.
+ * @param studyEndDate offering end date.
+ * @param referenceSubmission used for reference to the assessment date
+ * and possible auditing for the awards.
+ * @returns the saved assessment and its dependencies.
+ */
+async function saveIER12AssessmentFromTestInput(
+  db: E2EDataSources,
+  testInputAssessment: IER12Assessment,
+  assessment: StudentAssessment,
+  studyStartDate: string,
+  studyEndDate: string,
+  referenceSubmission: Date,
+): Promise<StudentAssessment> {
+  assessment.triggerType = testInputAssessment.triggerType;
+  // Simulates that the assessment date was produced one day after the reference date
+  // in case the assessmentDate was not provided. It helps creating different dates
+  // which allows a better data assertion.
+  assessment.assessmentDate =
+    testInputAssessment.assessmentDate ?? addDays(1, referenceSubmission);
+  assessment.assessmentData =
+    testInputAssessment.assessmentData as FullTimeAssessment;
+  assessment.workflowData = testInputAssessment.workflowData;
+  // Schedules and awards mapping.
+  const [firstDisbursement, secondDisbursement] =
+    assessment.disbursementSchedules;
+  const [firstDisbursementTestInput, secondDisbursementTestInput] =
+    testInputAssessment.disbursementSchedules;
+  mapTestInputToDisbursementAndAwards(
+    firstDisbursement,
+    firstDisbursementTestInput,
+    studyStartDate,
+    referenceSubmission,
+  );
+  if (secondDisbursement) {
+    // Calculate the middle date between offering start/end dates.
+    const midOfferingDateDays =
+      dateDifference(studyEndDate, studyStartDate) / 2;
```

Comment:
Just to check, normally if the study end date is in the past during the application submission, then we take today as the disbursement date right? And there will be only one disbursement even if the study weeks are more than 17 weeks right? I know, for the test the default program year is `2000`, but here the plan is not to make anything complicated and to follow a happy path scenario right?

Reply:
The study start/end dates do not consider that logic here. The idea here was just to produce some sort of different dates. While doing it the idea was to try to replicate a bit what the real dates would be. For the sake of validating the IER12, unless I am missing something, I would not go further trying to replicate the disbursements schedules logic.